### PR TITLE
Frg show token transfers in transactions

### PR DESCRIPTION
**Pull request by feliperenan**
*[resolves](https://github.com/poanetwork/poa-explorer/issues/75)*

## Motivation
We need to include on Address' Page in the transactions' tab the token transfers that addresses have exchanged.

## How token transfers will be shown in POA Explorer?
We'll have different scenarios according to the Address's Page. As an example, I going to use a real
token transfers operation that happened in Kovan network so that Guilherme Fre has transferred some freh coins (the token) to Amanda. You can see how the transaction looks like on [Etherscan](https://kovan.etherscan.io/tx/0xce242a2a20520c38305641e2ad15e42096ca4cd26017b2b7cf4d37d319867e5a)

Following the example above, in POA Explorer we'll have:
- On Guilherme Fre address' page we'll see a token transfer to Amanda with the label `OUT`.
- On Amanda address' page we'll see a token transfer from Guilherme Fre with the label `IN`.
- On Contract address' page we'll see a token transfer from Guilherme Fre to Amanda with the label `IN` since this contract was called to create the transaction.

## Screenshots
*uploading...*

## Technical things
In order to this work, we need to have Tokens and Token Transfers schemas. I'm creating these ones in this PR [based on it](https://github.com/poanetwork/poa-explorer/pull/334/files#diff-53ceb39a27d8b44fe37e2d869918470cR1) since we need these tables to build the query.

Talking about the query, I think that worth explains a business rule. Using the example of token transfers that I used above, Amanda won't have a transaction. But, why? 🤔 Because when a token transfer happens, in the transaction that represents it, the from_address will be who is transferring (fre_address) and the to_address will be the contract that created the token. Who is receiving the token (Amanda) will be in the token transfer as the to_address.

Since Amanda doesn't have a transaction and we need to show in her address' page that she received token. When getting all transactions given an address, we need to join the transactions' table together token transfers' table considering that the address also can be in `to_address_hash` from token transfers' table.

## How to test it?
While we don't have Token and Token Transfers indexed from blockchain in our database, we can use [this script](https://gist.github.com/feliperenan/b7aacba41bb54b7e045ff0a1032c8113) that I have created to populate our database.

## TODO 
- [ ] Front-end
- [ ] Feature tests

### CODE REVIEW

File: apps/explorer/lib/explorer/chain.ex [hunk 1]
```diff
@@ -2284,6 +2284,7 @@ defmodule Explorer.Chain do
     })
     |> order_by([transaction], desc: transaction.block_number, desc: transaction.index)
     |> handle_paging_options(paging_options)
+    |> preload(token_transfers: :token)
```
Comment (igorffs) [inline apps/explorer/lib/explorer/chain.ex]: This `fetch_transaction` function is being used by at least other 3 functions, that could impact in other use cases that don't need this preload. Can you confirm that? Maybe we should consider writing a new function for this. 

wdyt?

File: apps/explorer/lib/explorer/chain.ex [hunk 2]
```diff
@@ -2725,7 +2726,16 @@ defmodule Explorer.Chain do
             LIMIT 1)
           ],
              t.hash
-           ))
+           )) or
```
Comment (igorffs) [inline apps/explorer/lib/explorer/chain.ex]: This https://github.com/poanetwork/poa-explorer/pull/417/files#r203453837 might be a concern here too.

File: apps/explorer/priv/repo/migrations/20180713191837_create_tokens.exs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+defmodule Explorer.Repo.Migrations.CreateTokens do
```
Comment (igorffs) [inline apps/explorer/priv/repo/migrations/20180713191837_create_tokens.exs]: @alexgaribay these migrations came from #334.  Would you say there modeling is final? 

I'd like to create PR and merge them to master so all of our next PR's can be also merged. That would help us improve our workflow since we wouldn't need to rebase our PR's in a WIP. wdyt?

Comment (igorffs) [reply apps/explorer/priv/repo/migrations/20180713191837_create_tokens.exs]: I've opened a PR for this https://github.com/poanetwork/poa-explorer/pull/426. @feliperenan do you want to merge this one with it?

File: apps/explorer/lib/explorer/chain.ex (hunk 1 above)
Comment (feliperenan) [reply apps/explorer/lib/explorer/chain.ex]: It makes senses. I gonna split it to another function 👍

File: apps/explorer/lib/explorer/chain.ex [hunk 4]
```diff
@@ -2729,6 +2731,25 @@ defmodule Explorer.Chain do
     )
   end
 
+  defp where_exists_token_transfer_with_matching_address_hash(%Ecto.Query{from: {_table, Transaction}} = query, address_hash) do
```
Comment (josevalim) [inline apps/explorer/lib/explorer/chain.ex]: I believe this should be moved to the `Transaction` module. I have talked about this here: #423

Also, please note that the `Ecto.Query` struct is **private**. You should not access its fields. For example, this code will be broken in the upcoming Ecto 3.0.

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 5]
```diff
@@ -326,4 +339,38 @@ defmodule Explorer.Chain.Transaction do
...
+          WHERE tt."transaction_hash" = ? AND tt."to_address_hash" = ?
+        )
+        """,
+        unquote(transaction_hash),
+        unquote(bytes)
+      )
+    end
+  end
+
+  def where_address_fields_match(query, address_hash) do
```
Comment (josevalim) [inline apps/explorer/lib/explorer/chain/transaction.ex]: Missing docs maybe?

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 6]
```diff
@@ -326,4 +339,38 @@ defmodule Explorer.Chain.Transaction do
...
+    where(
+      query,
+      [t],
+      t.to_address_hash == ^address_hash or t.from_address_hash == ^address_hash or
+        (is_nil(t.to_address_hash) and
+           ^address_hash.bytes in fragment(
```
Comment (josevalim) [inline apps/explorer/lib/explorer/chain/transaction.ex]: Since we are here, maybe we should refactor this? I believe this could also be an exists query, similar to the one above.

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 7]
```diff
@@ -326,4 +339,55 @@ defmodule Explorer.Chain.Transaction do
...
+
+  def where_address_fields_match(query, address_hash, :to) do
+    where(query, [t], t.to_address_hash == ^address_hash)
+  end
+
+  def where_address_fields_match(query, address_hash, :from) do
+    where(query, [t], t.from_address_hash == ^address_hash)
+  end
+
+  def where_address_fields_match(%Ecto.Query{from: {_table, InternalTransaction}} = query, address_hash, nil) do
```
Comment (feliperenan) [inline apps/explorer/lib/explorer/chain/transaction.ex]: @josevalim As we won't support this kind of thing in Ecto 3, which is the better way to pattern match it?

Comment (josevalim) [reply apps/explorer/lib/explorer/chain/transaction.ex]: Don't pattern match on it, pass this as an explicit argument when invoking this function. :)

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 8]
```diff
@@ -326,4 +339,55 @@ defmodule Explorer.Chain.Transaction do
...
+          WHERE tt."transaction_hash" = ? AND tt."to_address_hash" = ?
+          )
+        ],
+        unquote(transaction_hash),
+        unquote(bytes)
+      )
+    end
+  end
+
+  def where_address_fields_match(query, address_hash, :to) do
```
Comment (josevalim) [inline apps/explorer/lib/explorer/chain/transaction.ex]: Do not forget the docs. :)

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 9]
```diff
@@ -326,4 +339,88 @@ defmodule Explorer.Chain.Transaction do
...
+  """
+  def where_address_fields_match(query, address_hash, :from) do
+    where(query, [t], t.from_address_hash == ^address_hash)
+  end
+
+  @doc """
+  Adds to the given internal_transaction's query a `where` condition considering that the given
+  addres_hash can be: to_address, from_address or created_contract_address_hash.
+  """
+  def where_address_fields_match(
```
Comment (tmecklem) [inline apps/explorer/lib/explorer/chain/transaction.ex]: I'm not sure this belongs here. It's matching on InternalTransaction, not Transaction. This gnarly Ecto query internal match will go away when we pull the created_contract_address_hash into the Transaction (https://github.com/poanetwork/poa-explorer/issues/430).

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 10]
```diff
@@ -326,4 +339,88 @@ defmodule Explorer.Chain.Transaction do
...
+          WHERE i."transaction_hash" = ? AND i."type" = 'create' AND i."created_contract_address_hash" = ?
+          )
+        ],
+        unquote(transaction_hash),
+        unquote(bytes)
+      )
+    end
+  end
+
+  @doc """
```
Comment (tmecklem) [inline apps/explorer/lib/explorer/chain/transaction.ex]: All of the `where_address_fields_match` functions operate either on the `InternalTransaction` or on the `Transaction` level, so they're shared between the two schema/structs. It seems like there's a missing module or level of abstraction here, or we should duplicate the `where_address_fields_match` logic in `InternalTransaction`.

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 11]
```diff
@@ -326,4 +339,88 @@ defmodule Explorer.Chain.Transaction do
...
+  Adds to the given transaction's query a `where` condition considering if the given address_hash
+  is equal to to_address_hash.
+  """
```
Comment (josevalim) [inline apps/explorer/lib/explorer/chain/transaction.ex]: @feliperenan this is a single function with multiple clauses. Only one `@doc` instance is going to prevail. Elixir v1.7 even warns in cases like this to avoid confusion. :) So you need to have only a single `@doc` declaration for all of them. :)

File: apps/explorer/lib/explorer/chain/transaction.ex (hunk 10 above)
Comment (josevalim) [reply apps/explorer/lib/explorer/chain/transaction.ex]: Good catch! In my opinion, we should go with the duplication from now. From my understanding of the code, only the first two clauses are shared and they are too simple. I would also say that the fact we tried to unify them is what led to some of the complexity we have right now, for example, in regards to accessing `Ecto.Query`'s `:from` field, which is private.

Comment (feliperenan) [reply apps/explorer/lib/explorer/chain/transaction.ex]: I opted for duplicating these function in `InternalTransaction` schema. Could you take a look?

Comment (josevalim) [reply apps/explorer/lib/explorer/chain/transaction.ex]: Beautiful.

File: apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex [hunk 12]
```diff
@@ -45,5 +45,21 @@
...
+    <%= for token_transfer <- @transaction.token_transfers do %>
+      <div class="offset-md-2 col-md-7 col-lg-8 d-flex flex-column">
+        <p class="tile-title mt-4"><%= gettext "Transfers" %></p>
+
+        <%= render(
+              address_template(@address, token_transfer),
+              from_address_hash: token_transfer.from_address_hash,
+              to_address_hash: token_transfer.to_address_hash,
+              locale: @locale
+            ) %>
```
Comment (jimmay5469) [inline apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex]: This was a bit confusing to me at first, I didn't realize that these partials included `[hash]->[hash]` rather than a single address.

I think I would prefer to not see the scenario combinations as their own partials, but rather, a partial used for the `to_address` and a partial used for the `from_address` with the `&rarr;` in this template itself, similar to what we've done above in this same file. If you would prefer, we could introduce a view helper like you have here, but to determine whether to display `_link.html` or `_responsive_hash.html`, that could possibly be an improvement on what we've been doing thus far as we wouldn't need the conditional in the template like we have above.

File: apps/explorer_web/lib/explorer_web/views/address_transaction_view.ex [hunk 13]
```diff
@@ -10,4 +10,12 @@ defmodule ExplorerWeb.AddressTransactionView do
       _ -> gettext("All")
     end
   end
+
+  def address_template(address, token_transfer) do
+    cond do
+      address.hash == token_transfer.from_address_hash -> "_from_address.html"
+      address.hash == token_transfer.to_address_hash -> "_to_address.html"
+      contract?(address) -> "_contract_address.html"
```
Comment (jimmay5469) [inline apps/explorer_web/lib/explorer_web/views/address_transaction_view.ex]: Is there a scenario where none of these conditions wouldn't be met? Are we using this as a means to filter out token transfers that are not to or from the current address or dealing with a contract?

Comment (feliperenan) [reply apps/explorer_web/lib/explorer_web/views/address_transaction_view.ex]: > Is there a scenario where none of these conditions wouldn't be met?

No. Since we are on Address's details page, we ever will have an address and in this section, it will be: `to`, `from` or a `contract`. 

> Are we using this as a means to filter out token transfers that are not to or from the current address or dealing with a contract?

We are dealing with addresses that are not `to` or `from`. I think that we can be more generic here and use `true -> _contract_address.html`

File: apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex (hunk 12 above)
Comment (feliperenan) [reply apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex]: @jimmay5469 Wdty about this version https://github.com/poanetwork/poa-explorer/pull/417/commits/9d74d5b9499a24b2cb7b9f2b32bb5e52286565d6?

File: apps/explorer/priv/repo/migrations/20180722143219_add_token_id_to_token_transfers.exs [hunk 14]
```diff
@@ -0,0 +1,9 @@
+defmodule Explorer.Repo.Migrations.AddTokenIdToTokenTransfers do
+  use Ecto.Migration
+
+  def change do
+    alter table(:token_transfers) do
+      add(:token_id, references(:tokens), null: false)
```
Comment (feliperenan) [inline apps/explorer/priv/repo/migrations/20180722143219_add_token_id_to_token_transfers.exs]: @alexgaribay We need to have the token reference from `token_transfers` to get token information from token transfers. Have you thought about doing that in any other way? 🤔

Comment (alexgaribay) [reply apps/explorer/priv/repo/migrations/20180722143219_add_token_id_to_token_transfers.exs]: That's part of the old schema. Master doesn't have it that way. The token transfer will have the contract address hash instead.

Comment (feliperenan) [reply apps/explorer/priv/repo/migrations/20180722143219_add_token_id_to_token_transfers.exs]: Since we have to get token's data from token transfer, what do you think about:

```elixir
# explorer/chain/address.ex 
has_one(:token, Token, foreign_key: :contract_address_hash)

# explorer/chain/token_transfer.ex
has_one(:token, through: [:token_contract_address, :token])
```

?

**Comment by feliperenan** [comment]
@jimmay5469 @tmecklem I have applied your suggestions, could you please take a look?

File: apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex [hunk 15]
```diff
@@ -46,5 +46,31 @@
...
+          </span>
+          &rarr;
+          <span data-test="token_transfer_address_hash">
+            <%= ExplorerWeb.AddressView.display_address_hash(@address, token_transfer.to_address, @locale) %>
+          </span>
+        </span>
+      </div>
+      <div class="col-sm-3 col-lg-2 d-flex flex-row flex-sm-column align-items-end">
+        <span class="tile-title">
+          <%= token_transfer.amount %> <%= token_transfer.token.symbol %>
```
Comment (feliperenan) [inline apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex]: @alexgaribay how you are storing the amount of tokens? For example, suppose that we have a token with 18 decimal places, if I transfer `100` tokens to another address, will we store `100` or `100000000000000000000`?

Comment (alexgaribay) [reply apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex]: @feliperenan Indexing token transfers just stores the raw value that is stored in the log, if available. Token information isn't available at the time of the main indexer where token transfers are index. Tokens are fetched asynchronously.

For displaying values, we'll have to treat it similar to how values are stored in Wei. The frontend will need to do some conversions to display values appropriately.

**TLDR** `1000000000000000000`

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 16]
```diff
@@ -4,7 +4,19 @@ defmodule Explorer.Chain.Transaction do
...
+  alias Explorer.Chain.{
+    Address,
+    Block,
+    Data,
+    Gas,
+    Hash,
+    InternalTransaction,
+    Log,
+    Wei,
+    TokenTransfer
```
Comment (Lucasnar) [inline apps/explorer/lib/explorer/chain/transaction.ex]: We were alphabetically sorting the aliases before. Do you know if that was intentional? If so, `TokenTransfer` should go before `Wei`. 🔡

Comment (feliperenan) [reply apps/explorer/lib/explorer/chain/transaction.ex]: I don't know haha, but I gonna change to follow this pattern 📔

File: apps/explorer_web/lib/explorer_web/views/address_transaction_view.ex [hunk 17]
```diff
@@ -28,4 +28,25 @@ defmodule ExplorerWeb.AddressTransactionView do
...
+  @doc """
+  Formats the given amount according to given decimals.
+
+  ## Examples
+
+  iex> formatted_token_amount(Decimal.new(2050000), 5)
+  "205"
+
+  iex> formatted_token_amount(Decimal.new(2050000), 4)
+  "2,050"
```
Comment (Lucasnar) [inline apps/explorer_web/lib/explorer_web/views/address_transaction_view.ex]: Shouldn't this be `205` and the above example `20.5`? 🤔

File: apps/explorer_web/lib/explorer_web/views/address_transaction_view.ex [hunk 18]
```diff
@@ -11,4 +11,35 @@ defmodule ExplorerWeb.AddressTransactionView do
...
+  @doc """
+  Formats the given amount according to given decimals.
+
+  ## Examples
+
+  iex> formatted_token_amount(Decimal.new(20500000), 5)
+  "205"
+
+  iex> formatted_token_amount(Decimal.new(205000000), 4)
+  "2,05"
```
Comment (Lucasnar) [inline apps/explorer_web/lib/explorer_web/views/address_transaction_view.ex]: Seeing a comma here in the return looks a little bit odd to me since Elixir will represent non-integer numbers with a dot (like in the next example of this file). Running this on `iex`, I get the following result:

```elixir
iex(2)> AddressTransactionView.formatted_token_amount(Decimal.new(205000000), 4)
"20500"
```

I don't know if you choose to not enable the doctests in this file, but in order to make sure those examples are working as expected, I would enable them in `address_transaction_view_test.exs` by adding the following line there:

```elixir
doctest ExplorerWeb.AddressTransactionView
```

You would also need to change the way you are referencing the function in order to run it (i.e. preppending `ExplorerWeb.AddressTransactionView.`).

File: apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex [hunk 19]
```diff
@@ -48,5 +54,31 @@
...
+
+    <%= for token_transfer <- @transaction.token_transfers do %>
+      <div class="offset-md-2 col-sm-7 col-lg-8 d-flex flex-column">
+        <span data-test="token_transfer">
+          <span data-test="token_transfer_address_hash">
+            <%= ExplorerWeb.AddressView.display_address_hash(@address, token_transfer.from_address, @locale) %>
+          </span>
+          &rarr;
+          <span data-test="token_transfer_address_hash">
+            <%= ExplorerWeb.AddressView.display_address_hash(@address, token_transfer.to_address, @locale) %>
```
Comment (jimmay5469) [inline apps/explorer_web/lib/explorer_web/templates/address_transaction/_transaction.html.eex]: Cleanup looks great! Thanks!
